### PR TITLE
remove bcrypt code from code-base (#4844)

### LINKED ISSUE #4844: do not compare credentials using bcrypt
Currently bcrypt is used to generate a hash from the secret key. This hash is used to compare the secret key send by clients during login. The whole bcrypt code is not necessary and uses the wrong primitive for this use case. A simple constant-time-compare is sufficient.

   
## Expected Behavior
Use constant time compare of secret key

## Current Behavior
Compute bcrypt value of secret key and use the equal function of bcrypt to compare the hash values.

## Possible Solution
Use subtle.ConstantTimeCompare

### PR DESCRIPTION
**Pull request by aead**
Bcrypt is not neccessary and not used properly. This change
replace the whole bcrypt hash computation through a constant time
compare and removes bcrypt from the code base.

Fixes #4844 

## How Has This Been Tested?
Manually

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by harshavardhana** [comment]
> hole bcrypt hash computation

what is a `hole bcrypt hash` ? do you mean `whole` ?

**Comment by aead** [comment]
Ahm, yes xD - I fix it!

**Review by balamurugana — COMMENTED** [review]
LGTM for removing codes

One question.
Does `subtle.ConstantTimeCompare()` prevent timing attack?

**Comment by aead** [comment]
Yes, that's exactly the function used in bcrypt for comparing the computed hash values. It compares two byte slices in constant time. This prevents timing attacks.

**Comment by aead** [comment]
Closed until we decide what to do...

**Comment by aead** [comment]
This PR should fulfill the requirements - reopened

File: cmd/credential.go [hunk 1]
```diff
@@ -80,64 +80,44 @@ func (cred credential) Equal(ccred credential) bool {
...
-		}
-
-		cred.secretKeyHash = secretKeyHash
+	if subtle.ConstantTimeCompare([]byte(cred.AccessKey), []byte(ccred.AccessKey)) != 1 {
+		return false
 	}
-
-	return (cred.AccessKey == ccred.AccessKey &&
-		bcrypt.CompareHashAndPassword(cred.secretKeyHash, []byte(ccred.SecretKey)) == nil)
+	return subtle.ConstantTimeCompare([]byte(cred.SecretKey), []byte(ccred.SecretKey)) == 1
```
Comment (balamurugana) [inline cmd/credential.go]: How about doing below?
```golang
return (cred.AccessKey == ccred.AccessKey && subtle.ConstantTimeCompare([]byte(cred.SecretKey), []byte(ccred.SecretKey)) == 1)
```

File: cmd/credential.go [hunk 2]
```diff
@@ -80,64 +80,44 @@ func (cred credential) Equal(ccred credential) bool {
...
 }
 
 // Initialize a new credential object
 func getNewCredential(accessKeyLen, secretKeyLen int) (cred credential, err error) {
-	// Generate access key.
-	keyBytes := make([]byte, accessKeyLen)
-	_, err = rand.Read(keyBytes)
-	if err != nil {
-		return cred, err
+	// Generate access key - the access key consits only of alpha-numeric digits.
```
Comment (balamurugana) [inline cmd/credential.go]: I think you can keep function unchanged to minimize this PR

File: cmd/credential.go (hunk 1 above)
Comment (aead) [reply cmd/credential.go]: The access key is public information so this is fine. However for me it looks a lite bit wired when credentials are compared and some parts are constant time and some aren't. It's just easier to miss something in the future - but the decision is up to you...

File: cmd/credential.go (hunk 2 above)
Comment (aead) [reply cmd/credential.go]: This PR also cleans up the code a bit - like unnecessary if-else. aso.

File: cmd/credential.go (hunk 1 above)
Comment (balamurugana) [reply cmd/credential.go]: I think you can do the change.

File: cmd/credential.go (hunk 2 above)
Comment (balamurugana) [reply cmd/credential.go]: Yep. Its better to send different PR doing cleanup.

File: cmd/credential.go (hunk 1 above)
Comment (aead) [reply cmd/credential.go]: done

File: cmd/credential.go (hunk 2 above)
Comment (aead) [reply cmd/credential.go]: done

**Review by harshavardhana — APPROVED** [review]

**Review by balamurugana — APPROVED** [review]

**Comment by harshavardhana** [comment]
Can you squash your commits? @aead

**Comment by aead** [comment]
Squashed

**Pull request merged**